src/app/cmd/flatten_layers.cpp

```diff
@@ -97,44 +127,89 @@ void FlattenLayers::onExecute()
     RestoreVisibleLayers restore;
     restore.showSelectedLayers(sprite, layers);
 
+    // Map draw area to image coords
+    const gfx::ClipF area_to_image(0, 0, area);
+
     // Copy all frames to the background.
     for (frame_t frame(0); frame<sprite->totalFrames(); ++frame) {
       // Clear the image and render this frame.
       clear_image(image.get(), bgcolor);
-      render.renderSprite(image.get(), sprite, frame);
+      render.renderSprite(image.get(), sprite, frame, area_to_image);
 
-      // TODO Keep cel links when possible
+      // Get exact bounds for rendered frame
+      gfx::Rect bounds = image->bounds();
+      const bool shrink = doc::algorithm::shrink_bounds(
+        image.get(), image->maskColor(), nullptr,
+        image->bounds(), bounds);
 
-      ImageRef cel_image;
+      // Skip when fully transparent
       Cel* cel = flatLayer->cel(frame);
+      if (!shrink) {
+        if (!newFlatLayer && cel)
+          executeAndAdd(new cmd::RemoveCel(cel));
+
+        continue;
+      }
+
+      // Apply shrunk bounds to new image
+      const ImageRef new_image(doc::crop_image(
+        image.get(), bounds, image->maskColor()));
+
+      // Replace image on existing cel
       if (cel) {
+
+        // TODO Keep cel links when possible
+
         if (cel->links())
           executeAndAdd(new cmd::UnlinkCel(cel));
 
-        cel_image = cel->imageRef();
+        const ImageRef cel_image = cel->imageRef();
         ASSERT(cel_image);
 
-        executeAndAdd(
-          new cmd::CopyRect(cel_image.get(), image.get(),
-                            gfx::Clip(0, 0, image->bounds())));
+        // Reset cel properties when flattening in-place
+        if (!newFlatLayer) {
+          executeAndAdd(new cmd::SetCelOpacity(cel, 255));
```

Comment:
We could create `SetCelOpacity()` only when it's required (this is something that wasn't present in the old impl of merge down but it'd be nice to have):
```suggestion
          if (cel->opacity() != 255)
            executeAndAdd(new cmd::SetCelOpacity(cel, 255));
```
One thing I noticed is the z-index of the resulting cel, I'm not sure if we should do something like:
```c++
          if (cel->zIndex() != newZIndex)
            executeAndAdd(new cmd::SetCelZIndex(cel, newZIndex));
```
The value of `newZIndex` could be just `0`, but there are some special cases that we could take care (I'll attach an example in the general comment).